### PR TITLE
[DSV4] Add native H16 CuTe sparse prefill backend

### PR DESCRIPTION
**Pull request by xuqianmamba**
## Motivation

DeepSeek-V4 has 128 global query heads, so tensor parallelism with TP=8 leaves
16 logical query heads per rank. The existing SM90 FlashMLA sparse-prefill path
operates on an H64-padded query.

This PR adds an explicit, experimental native-H16 CuTe DSL backend:

```bash
--dsv4-prefill-backend cutedsl_h16
```

The backend avoids computing and materializing the 48 padded query heads and
returns a compact `[TQ, 16, 512]` output. The default `auto` behavior is
unchanged.

## Modifications

- Add an SM90 BF16 native-H16 CuTe DSL sparse-prefill kernel using an `m64n16`
  WGMMA decomposition and online softmax.
- Add `cutedsl_h16` to the DeepSeek-V4 sparse-prefill backend selection.
- Validate the supported launch contract:
  - CUDA SM90;
  - BF16 query and sparse workspace;
  - exactly 16 TP-local logical query heads;
  - Q/K/V width 512;
  - combined TOPK aligned to 128 and no greater than 8192.
- Use a configuration-sized aligned C128 TOPK bucket so changing live chunk
  lengths does not introduce a new JIT specialization.
- Cache the padded combined-index representation for reuse across layers.
- Keep decode, speculative, TBO, non-eager EXTEND, and CUDA graph/capture
  forwards on their existing paths.
- Reject explicit `cutedsl_h16` selection on unsupported HIP/NPU platforms
  instead of silently ignoring it.
- Add SM90 kernel tests, independent oracle/parity coverage, dispatch and CLI
  tests, a three-provider CUDA benchmark, and user documentation.

## Accuracy Tests

Validation used source commit
`5d0e688a3ca49bb521be4e20e835d415bb4f72af` on 8x NVIDIA H20-3e (SM90), with
CUDA 13.0, PyTorch 2.13.0+cu130, Triton 3.7.1,
`nvidia-cutlass-dsl==4.6.2`, `sglang-kernel==0.4.6.post1`,
`sgl-deep-gemm==0.1.5.post3`, and FlashInfer 0.6.17. A later docs-only commit
makes the launch examples copy/pasteable and does not change the tested code.

### Kernel and integration tests

```text
FlashInfer DSV4 MXFP4 direct-parity preflight: 3 passed
Native-H16 SM90 GPU test suite:             15 passed
DSV4 dispatch and server-argument tests:    14 passed
```

For the RMS1 / TOPK=1152 parity fixture:

```text
max absolute error:  9.765625e-04
mean absolute error: 8.782039e-06
```

The GPU suite covers dynamic query and KV lengths, TOPK 128 through 8192,
invalid indices, poisoned row zero, H16/H64 storage parity, alignment
validation, repeated allocation, non-default CUDA streams, and online-rescale
stress cases.

### Full-model accuracy

Both variants used DeepSeek-V4-Pro with TP=8, context length 49,152, random
seed 1234, maximum 16 running requests, chunked prefill size 4096, maximum
16,384 prefill tokens, and identical server/evaluation arguments except for
the backend and port.

GSM8K used all 1,314 questions, 5-shot prompting, maximum 512 new tokens,
parallelism 32, and temperature 0.

| Backend | Accuracy | Invalid | Latency (s) | Output tok/s | Exit code |
|---|---:|---:|---:|---:|---:|
| `flashmla_sparse` | 0.951 | 0.000 | 505.030 | 236.364 | 0 |
| `cutedsl_h16` | 0.952 | 0.000 | 493.418 | 243.384 | 0 |

On this public full-model evaluation, H16 reduced latency by 2.3% and increased
output throughput by 3.0% while preserving accuracy.

The candidate server emitted 8
`DSV4_CUTEDSL_H16_SPARSE_PREFILL_HIT` markers, one from each TP rank,
confirming that the full-model workload reached the new backend.

### Sparse-prefill CUDA microbenchmark

Warmed public-API CUDA-event latency at TQ=4096 and SKV=49152:

| TOPK | FlashMLA BF16 H64 p50 (us) | FlashMLA Q8 H64 p50 (us) | CuTe BF16 H16 p50 (us) | H16 vs BF16 | H16 vs Q8 |
|---:|---:|---:|---:|---:|---:|
| 128 | 700.592 | 538.000 | 235.008 | 2.9811x | 2.2893x |
| 384 | 1676.160 | 1136.352 | 557.120 | 3.0086x | 2.0397x |
| 512 | 2165.616 | 1438.640 | 718.672 | 3.0134x | 2.0018x |
| 640 | 2655.984 | 1740.064 | 880.208 | 3.0175x | 1.9769x |
| 1152 | 4557.152 | 2913.088 | 1511.008 | 3.0160x | 1.9279x |

These numbers measure each provider's public attention call, not a
raw-kernel-only launch or end-to-end serving latency. FlashMLA BF16 and Q8
produce an H64 output plus max/LSE auxiliaries, while CuTe returns a compact H16
output. Q8 inputs are prequantized before timing; its query cast and DSV4 KV
gather/dequantize/requantize pipeline are excluded.

## Checklist

- [x] Format the code according to the repository configuration. The changed
  files pass the repository-pinned Black, isort, Ruff selection, codespell,
  whitespace, AST, and registered-test lint checks.
- [x] Add unit tests.
- [x] Update documentation.
- [x] Provide accuracy and speed benchmark results.
- [x] Follow the SGLang code-style guidance.

### CODE REVIEW

**Comment by xuqianmamba** [comment]
@BBuf @ispobock Could you please help add the run-ci label and review this PR when convenient? The required CI gate is currently blocked because the label is missing. Thanks!
